# IBM Data Science Capstone Project

## T Farrington

## Canberra Suburbs: Finding a place to live

##### 26th November 2019

### Part 0 - Initialise Notebook

In this section we import all the modules required for this project

#### Import modules

In [2]:
print('Importing modules ...')

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium
from geopy.geocoders import Nominatim
import geocoder
import requests

from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler, normalize, scale
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.decomposition import PCA
from sklearn.metrics import mean_squared_error, r2_score

print('Modules imported')

Importing modules ...
Modules imported


### Part 1 - Data Preparation 

#### Read Canberra suburbs CSV file into DataFrame

In [3]:
# Source file
canberra_csv = r'Canberra_Suburbs.csv'
# Read CSV into dataframe
df = pd.read_csv(canberra_csv)
# Sort by Suburb name (alph ASC)
df.sort_values(by=['Suburb'], inplace = True)
df.reset_index(drop=True, inplace = True)
df.head()

Suburb   Longitude   Latitude
0    Acton  149.112771 -35.281319
1  Ainslie  149.148232 -35.263600
2   Amaroo  149.127417 -35.168831
3   Aranda  149.080925 -35.257771
4    Banks  149.100662 -35.471861

##### Check dataframe

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 116 entries, 0 to 115
Data columns (total 3 columns):
Suburb       116 non-null object
Longitude    116 non-null float64
Latitude     116 non-null float64
dtypes: float64(2), object(1)
memory usage: 2.8+ KB


116 Suburbs  
no nulls  
Suburb name is type 'object'  
Longitude and Latitude are type 'float64'  
All as expected, no further cleaning required

#### Get Domain Suburb IDs

###### Domain api credentials

In [5]:
domain_key = 'XXXXXXXXXXXX'

##### Make first set of API calls

In [6]:
S_names = df['Suburb'].tolist()
i=0
responses = []
for name in S_names:
    url = r'https://api.domain.com.au/v1/addressLocators?api_key={}&searchLevel=Suburb&suburb={}&state=ACT'.format(
        domain_key, 
        name)
    
    response = requests.get(url).json()[0]["ids"][0]
    response["level"] = name
    responses.append(response)

suburbs = pd.DataFrame(responses)
suburbs.rename(columns={'level' : 'Suburb', 'id' : 'Domain ID'}, inplace = True)
suburbs = suburbs[['Suburb', 'Domain ID']]
suburbs.head()

Suburb  Domain ID
0    Acton         61
1  Ainslie         71
2   Amaroo         91
3   Aranda        101
4    Banks        131

##### Check Dataframe 

In [7]:
suburbs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 116 entries, 0 to 115
Data columns (total 2 columns):
Suburb       116 non-null object
Domain ID    116 non-null int64
dtypes: int64(1), object(1)
memory usage: 1.9+ KB


##### Merge base dataframes

In [8]:
canberra = pd.merge(df, suburbs, on=['Suburb'])
canberra = canberra[['Suburb', 'Domain ID', 'Longitude', 'Latitude']]
canberra.head(20)

Suburb  Domain ID   Longitude   Latitude
0              Acton         61  149.112771 -35.281319
1            Ainslie         71  149.148232 -35.263600
2             Amaroo         91  149.127417 -35.168831
3             Aranda        101  149.080925 -35.257771
4              Banks        131  149.100662 -35.471861
5             Barton        141  149.137112 -35.307921
6              Beard       6451  149.211188 -35.342198
7          Belconnen        171  149.068356 -35.235251
8             Bonner       3041  149.142229 -35.157298
9           Bonython        191  149.077786 -35.434445
10           Braddon        221  149.136733 -35.271422
11             Bruce        241  149.093361 -35.245587
12           Calwell        271  149.106573 -35.442859
13          Campbell        291  149.154939 -35.288914
14  Canberra Airport       6461  149.194880 -35.303439
15      Capital Hill        321  149.124435 -35.308179
16             Casey       3291  149.094391 -35.166508
17           Chapman        351  149.036772 -35.354587
18         Charnwood        361  149.033784 -35.200478
19           Chifley        371  149.078003 -35.353877

### Get housing data

#### Get median house prices for last 5 years

In [9]:
suburb_ids = canberra['Domain ID'].tolist()
responses = []
median_price_2014 = []
median_price_2019 = []
growth = []

for suburb_id in suburb_ids:
    url = 'https://api.domain.com.au/v1/suburbPerformanceStatistics?api_key={}&state=ACT&suburbId={}&propertyCategory=house&chronologicalSpan=12&tPlusFrom=1&tPlusTo=6&values=MedianSoldPrice'.format(
        domain_key,
        suburb_id)
    try:
        responses.append(requests.get(url).json())
    except:
        responses.append(np.nan)

In [10]:
for response in responses:
    if str(response).startswith('{'):
        # Extract -5yr prices
        price2014 = response["series"]['seriesInfo'][0]['values']['medianSoldPrice']
        if price2014 == None:
            median_price_2014.append(np.nan)
        else:
            median_price_2014.append(price2014)
        # Extract current prices 
        price2019 = response["series"]['seriesInfo'][-1]['values']['medianSoldPrice']
        if price2019 == None:
            median_price_2019.append(np.nan)
        else:
            median_price_2019.append(price2019)
        # Calculate growth
        if price2014 == None or price2014 == 0 or price2019 == None:
            growth.append(np.nan)
        else:
            growth.append(round(float(price2019)/float(price2014) * 100, 2))
    else:
        median_price_2014.append(np.nan)
        median_price_2019.append(np.nan)
        growth.append(np.nan)

In [11]:
# Build result dataframe
house_data = list(zip(suburb_ids, median_price_2014, median_price_2019, growth))
house_data
canberra_houses = pd.DataFrame(house_data, columns = ['Domain ID', 'Median Price 2014', 'Median Price 2019', '5 yr Growth (%)'])
canberra_houses.head(20)

Domain ID  Median Price 2014  Median Price 2019  5 yr Growth (%)
0          61                NaN                NaN              NaN
1          71           695000.0           975000.0           140.29
2          91           544000.0           687000.0           126.29
3         101           662000.0           960000.0           145.02
4         131           470000.0           485000.0           103.19
5         141                NaN                NaN              NaN
6        6451                NaN                NaN              NaN
7         171                NaN           420000.0              NaN
8        3041           533000.0           661000.0           124.02
9         191           528000.0           643000.0           121.78
10        221                NaN          1250000.0              NaN
11        241           730000.0           750000.0           102.74
12        271           465000.0           589000.0           126.67
13        291           843000.0          1255000.0           148.87
14       6461                NaN                NaN              NaN
15        321                NaN                NaN              NaN
16       3291           433000.0           600000.0           138.57
17        351           715000.0           908000.0           126.99
18        361           402000.0           478000.0           118.91
19        371           585000.0           791000.0           135.21

#### Merge dataframes

In [12]:
canberra = pd.merge(canberra, canberra_houses, on=['Domain ID'])
canberra.head(20)

Suburb  Domain ID   Longitude   Latitude  Median Price 2014  \
0              Acton         61  149.112771 -35.281319                NaN   
1            Ainslie         71  149.148232 -35.263600           695000.0   
2             Amaroo         91  149.127417 -35.168831           544000.0   
3             Aranda        101  149.080925 -35.257771           662000.0   
4              Banks        131  149.100662 -35.471861           470000.0   
5             Barton        141  149.137112 -35.307921                NaN   
6              Beard       6451  149.211188 -35.342198                NaN   
7          Belconnen        171  149.068356 -35.235251                NaN   
8             Bonner       3041  149.142229 -35.157298           533000.0   
9           Bonython        191  149.077786 -35.434445           528000.0   
10           Braddon        221  149.136733 -35.271422                NaN   
11             Bruce        241  149.093361 -35.245587           730000.0   
12           Calwell        271  149.106573 -35.442859           465000.0   
13          Campbell        291  149.154939 -35.288914           843000.0   
14  Canberra Airport       6461  149.194880 -35.303439                NaN   
15      Capital Hill        321  149.124435 -35.308179                NaN   
16             Casey       3291  149.094391 -35.166508           433000.0   
17           Chapman        351  149.036772 -35.354587           715000.0   
18         Charnwood        361  149.033784 -35.200478           402000.0   
19           Chifley        371  149.078003 -35.353877           585000.0   

    Median Price 2019  5 yr Growth (%)  
0                 NaN              NaN  
1            975000.0           140.29  
2            687000.0           126.29  
3            960000.0           145.02  
4            485000.0           103.19  
5                 NaN              NaN  
6                 NaN              NaN  
7            420000.0              NaN  
8            661000.0           124.02  
9            643000.0           121.78  
10          1250000.0              NaN  
11           750000.0           102.74  
12           589000.0           126.67  
13          1255000.0           148.87  
14                NaN              NaN  
15                NaN              NaN  
16           600000.0           138.57  
17           908000.0           126.99  
18           478000.0           118.91  
19           791000.0           135.21

In [13]:
canberra.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 116 entries, 0 to 115
Data columns (total 7 columns):
Suburb               116 non-null object
Domain ID            116 non-null int64
Longitude            116 non-null float64
Latitude             116 non-null float64
Median Price 2014    87 non-null float64
Median Price 2019    96 non-null float64
5 yr Growth (%)      86 non-null float64
dtypes: float64(5), int64(1), object(1)
memory usage: 7.2+ KB


In [14]:
canberra.describe()

Domain ID   Longitude    Latitude  Median Price 2014  \
count   116.000000  116.000000  116.000000       8.700000e+01   
mean   1575.568966  149.097883  -35.296502       6.050575e+05   
std    1905.070759    0.047611    0.083825       2.042642e+05   
min      61.000000  148.925092  -35.508979       4.020000e+05   
25%     618.500000  149.064579  -35.352567       4.815000e+05   
50%    1036.000000  149.100564  -35.297153       5.440000e+05   
75%    1531.000000  149.133244  -35.226085       6.550000e+05   
max    7841.000000  149.227164  -35.153357       1.900000e+06   

       Median Price 2019  5 yr Growth (%)  
count       9.600000e+01        86.000000  
mean        7.879167e+05       128.451047  
std         2.865158e+05        10.593759  
min         4.200000e+05       101.540000  
25%         6.000000e+05       121.817500  
50%         7.140000e+05       128.640000  
75%         8.697500e+05       134.620000  
max         2.210000e+06       154.390000

In [31]:
canberra_sorted = canberra[(canberra['Median Price 2019'].between((canberra['Median Price 2019'].mean() * 0.8), (canberra['Median Price 2019'].mean() * 1.2))) & (canberra['5 yr Growth (%)'] >= canberra['5 yr Growth (%)'].mean())] 
#canberra_sorted['5 yr Growth (%)'].fillna(100, inplace = True)
canberra_house = canberra_sorted.sort_values(['5 yr Growth (%)', 'Median Price 2019'], ascending = [False, True])
canberra_house.head(10)

Suburb  Domain ID   Longitude   Latitude  Median Price 2014  \
114       Wright       7121  149.033242 -35.320674           580000.0   
23          Cook        441  149.066321 -35.260416           529000.0   
42        Fraser        661  149.045276 -35.191903           500000.0   
74        Mawson       1141  149.100467 -35.363007           545000.0   
81   Narrabundah       1211  149.148882 -35.335116           660000.0   
31         Duffy        551  149.033466 -35.334721           545000.0   
87         Oxley       1331  149.078932 -35.409159           476000.0   
41      Franklin       2301  149.143494 -35.197892           530000.0   
64        Kaleen        951  149.108439 -35.226296           530000.0   
52       Hackett        761  149.162325 -35.250558           663000.0   

     Median Price 2019  5 yr Growth (%)  
114           880000.0           151.72  
23            765000.0           144.61  
42            723000.0           144.60  
74            760000.0           139.45  
81            920000.0           139.39  
31            759000.0           139.27  
87            655000.0           137.61  
41            725000.0           136.79  
64            725000.0           136.79  
52            899000.0           135.60

### Suburb amenities

For this part the FourSquare venue api will be used.  

The returned venue data will then be used to build a profile of each suburb. 

##### Foursquare initialisation

FourSquare credentials

In [32]:
client_id = 'XXXXXXXXXXXX'
client_secret = 'XXXXXXXXXXXX'
version = 20180605

##### Define venue function

In [38]:
def getVenues(suburbs, lats, longs, radius = 1000, limit = 100):
    '''
    return up to 100 venues for each suburb
    limited by radius from suburb centroid
    and category list
    '''
    # initialise list
    venues_list = []
    # loop through suburbs
    for suburb, lat, long in zip(suburbs, lats, longs):
        
        print(f'Searching {suburb} ...', end="\r")
        
        # create query
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            client_id,
            client_secret,
            version,
            lat,
            long,
            radius,
            #category_list,
            limit)
        
        # execute query
        try:
            results = requests.get(url).json()["response"]['groups'][0]['items']
            #print(results)
        
            # clean query response
            venues_list.append([(
                suburb, 
                lat, 
                long, 
                v['venue']['name'], 
                v['venue']['location']['lat'], 
                v['venue']['location']['lng'],  
                v['venue']['categories'][0]['name']) for v in results])
        except:
            print(f'No results for {suburb}')
        
    # create dataframe
    venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    venues.columns = ['Suburb', 
                  'Suburb Latitude', 
                  'Suburb Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    return venues    

#### Search Canberra by suburb

In [39]:
#category_list = ['4bf58dd8d48988d1f9941735',

canberra_venues = getVenues(suburbs=canberra['Suburb'],
                          lats=canberra['Latitude'],
                          longs=canberra['Longitude'],
                          #category_list=[]
                          radius = 1000)
print(canberra_venues.shape)
canberra_venues.head()

(1448, 7) Yarralumla ...e ....


Suburb  Suburb Latitude  Suburb Longitude  \
0  Acton       -35.281319        149.112771   
1  Acton       -35.281319        149.112771   
2  Acton       -35.281319        149.112771   
3  Acton       -35.281319        149.112771   
4  Acton       -35.281319        149.112771   

                                 Venue  Venue Latitude  Venue Longitude  \
0  Australian National Botanic Gardens      -35.278050       149.109428   
1                       Llewellyn Hall      -35.280604       149.123442   
2        National Film & Sound Archive      -35.283131       149.121143   
3              Monster Kitchen and Bar      -35.285122       149.122547   
4                             BrodDogs      -35.278428       149.122443   

     Venue Category  
0  Botanical Garden  
1      Concert Hall  
2            Museum  
3         Hotel Bar  
4        Food Truck

#### Analysing each suburb

In [40]:
print('There are {} unique venue categories in Canberra.'.format(len(canberra_venues['Venue Category'].unique())))

# one-hot-encoding venue categories
canberra_encoded = pd.get_dummies(canberra_venues[['Venue Category']], prefix="", prefix_sep="")
# add postcodes back
canberra_encoded['Suburb'] = canberra_venues['Suburb']
# and reorganise column order
fixed_columns = [canberra_encoded.columns[-1]] + list(canberra_encoded.columns[:-1])
canberra_encoded = canberra_encoded[fixed_columns]
print(f'Encoded dataframe shape: {canberra_encoded.shape}')
# aggregate encoded venues by postcode
canberra_grouped = canberra_encoded.groupby('Suburb').mean().reset_index()
print(f'Grouped dataframe shape: {canberra_grouped.shape}')
canberra_grouped

There are 207 unique venue categories in Canberra.
Encoded dataframe shape: (1448, 208)
Grouped dataframe shape: (111, 208)


Suburb   Airport  Airport Lounge  Airport Terminal  \
0               Acton  0.000000        0.000000               0.0   
1             Ainslie  0.000000        0.000000               0.0   
2              Amaroo  0.000000        0.000000               0.0   
3              Aranda  0.000000        0.000000               0.0   
4               Banks  0.000000        0.000000               0.0   
5              Barton  0.000000        0.000000               0.0   
6               Beard  0.000000        0.000000               0.0   
7           Belconnen  0.000000        0.000000               0.0   
8              Bonner  0.000000        0.000000               0.0   
9            Bonython  0.000000        0.000000               0.0   
10            Braddon  0.000000        0.000000               0.0   
11              Bruce  0.000000        0.000000               0.0   
12            Calwell  0.000000        0.000000               0.0   
13           Campbell  0.000000        0.000000               0.0   
14   Canberra Airport  0.058824        0.058824               0.0   
15       Capital Hill  0.000000        0.000000               0.0   
16              Casey  0.000000        0.000000               0.0   
17            Chapman  0.000000        0.000000               0.0   
18          Charnwood  0.000000        0.000000               0.0   
19            Chifley  0.000000        0.000000               0.0   
20           Chisholm  0.000000        0.000000               0.0   
21               City  0.000000        0.000000               0.0   
22             Conder  0.000000        0.000000               0.0   
23               Cook  0.000000        0.000000               0.0   
24             Coombs  0.000000        0.000000               0.0   
25              Crace  0.000000        0.000000               0.0   
26             Curtin  0.000000        0.000000               0.0   
27             Deakin  0.000000        0.000000               0.0   
28    Denman Prospect  0.000000        0.000000               0.0   
29            Dickson  0.000000        0.000000               0.0   
..                ...       ...             ...               ...   
81           Nicholls  0.000000        0.000000               0.0   
82           O'Connor  0.000000        0.000000               0.0   
83        Oaks Estate  0.000000        0.000000               0.0   
84              Oxley  0.000000        0.000000               0.0   
85               Page  0.000000        0.000000               0.0   
86         Palmerston  0.000000        0.000000               0.0   
87             Parkes  0.000000        0.000000               0.0   
88             Pearce  0.000000        0.000000               0.0   
89            Phillip  0.000000        0.000000               0.0   
90           Red Hill  0.000000        0.000000               0.0   
91               Reid  0.000000        0.000000               0.0   
92         Richardson  0.000000        0.000000               0.0   
93             Rivett  0.000000        0.000000               0.0   
94            Russell  0.000000        0.000000               0.0   
95            Scullin  0.000000        0.000000               0.0   
96             Spence  0.000000        0.000000               0.0   
97           Stirling  0.000000        0.000000               0.0   
98          Symonston  0.000000        0.000000               0.0   
99             Tharwa  0.000000        0.000000               0.0   
100          Theodore  0.000000        0.000000               0.0   
101           Throsby  0.000000        0.000000               0.0   
102           Torrens  0.000000        0.000000               0.0   
103            Turner  0.000000        0.000000               0.0   
104         Wanniassa  0.000000        0.000000               0.0   
105         Waramanga  0.000000        0.000000               0.0   
106            Watson  0.000000        0.000000               0.0   
107        We

In [47]:
del suburb_venues

#### Top 10 venues per postcode

In [49]:
def return_most_common_venues(row, n):
    '''
    returns top n most frequent venues in descending order
    '''
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:n]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Suburb']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
suburb_venues = pd.DataFrame(columns=columns)
suburb_venues['Suburb'] = canberra_grouped['Suburb']

for ind in np.arange(canberra_grouped.shape[0]):
    suburb_venues.iloc[ind, 1:] = return_most_common_venues(canberra_grouped.iloc[ind, :], num_top_venues)

# view shape and head of postcode_venue dataframe
print(f'Suburb_venues shape: {suburb_venues.shape}')
suburb_venues.head()


Suburb_venues shape: (111, 11)


Suburb 1st Most Common Venue      2nd Most Common Venue  \
0    Acton            Food Truck                       Café   
1  Ainslie           Sports Club                       Café   
2   Amaroo           Supermarket          Indian Restaurant   
3   Aranda                  Café  Middle Eastern Restaurant   
4    Banks           Pizza Place       Gym / Fitness Center   

        3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                       River                Museum        Botanical Garden   
1              Shopping Plaza         Grocery Store                  Bakery   
2                        Lake         Grocery Store          Shopping Plaza   
3             Nature Preserve    Chinese Restaurant  Thrift / Vintage Store   
4  Construction & Landscaping         Grocery Store                   Trail   

  6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
0           Coffee Shop           Concert Hall        Science Museum   
1     Fish & Chips Shop  Australian Restaurant      Business Service   
2   Filipino Restaurant               Fountain      Football Stadium   
3           Yoga Studio    Fried Chicken Joint              Fountain   
4           Yoga Studio   Fast Food Restaurant      Football Stadium   

  9th Most Common Venue 10th Most Common Venue  
0                 Plaza         Sandwich Place  
1                 Hotel                    Pub  
2            Food Truck             Food Court  
3      Football Stadium             Food Truck  
4            Food Truck             Food Court

#### Categorise suburbs into clusters

In [50]:
# really should optimise for k but going with 10 for now
# number of clusters
k = 10

# prep dataframe for clustering
canberra_clustering = canberra_grouped.drop('Suburb', 1)
# set-up and fit k-means clustering
kmeans = KMeans(n_clusters=k, random_state=0).fit(canberra_clustering)

In [51]:
#del canberra_suburbs
# add cluster labels into suburb_venues dataframe
suburb_venues.insert(0, 'Cluster Labels', kmeans.labels_)
# merge canberra_grouped with canberra_data to add latitude/longitude for each neighborhood
canberra_suburbs = pd.merge(canberra, suburb_venues, on=['Suburb'])
#canberra_suburbs = canberra_suburbs.join(suburb_venues.set_index('Suburb'), on='Suburb')
#canberra_suburbs = canberra_suburbs.join(suburb_venues, on='Suburb')
suburb_venues.dropna(inplace = True)

In [52]:
# check results
canberra_suburbs

Suburb  Domain ID   Longitude   Latitude  Median Price 2014  \
0               Acton         61  149.112771 -35.281319                NaN   
1             Ainslie         71  149.148232 -35.263600           695000.0   
2              Amaroo         91  149.127417 -35.168831           544000.0   
3              Aranda        101  149.080925 -35.257771           662000.0   
4               Banks        131  149.100662 -35.471861           470000.0   
5              Barton        141  149.137112 -35.307921                NaN   
6               Beard       6451  149.211188 -35.342198                NaN   
7           Belconnen        171  149.068356 -35.235251                NaN   
8              Bonner       3041  149.142229 -35.157298           533000.0   
9            Bonython        191  149.077786 -35.434445           528000.0   
10            Braddon        221  149.136733 -35.271422                NaN   
11              Bruce        241  149.093361 -35.245587           730000.0   
12            Calwell        271  149.106573 -35.442859           465000.0   
13           Campbell        291  149.154939 -35.288914           843000.0   
14   Canberra Airport       6461  149.194880 -35.303439                NaN   
15       Capital Hill        321  149.124435 -35.308179                NaN   
16              Casey       3291  149.094391 -35.166508           433000.0   
17            Chapman        351  149.036772 -35.354587           715000.0   
18          Charnwood        361  149.033784 -35.200478           402000.0   
19            Chifley        371  149.078003 -35.353877           585000.0   
20           Chisholm        381  149.122858 -35.421828           445000.0   
21               City        401  149.128841 -35.281305                NaN   
22             Conder        431  149.100971 -35.457690           530000.0   
23               Cook        441  149.066321 -35.260416           529000.0   
24             Coombs       7161  149.041985 -35.316071                NaN   
25              Crace       3781  149.107130 -35.203229           620000.0   
26             Curtin        501  149.075315 -35.325451           690000.0   
27             Deakin        521  149.102438 -35.318825           895000.0   
28    Denman Prospect       7821  149.024625 -35.294279                NaN   
29            Dickson        531  149.140231 -35.254024           650000.0   
..                ...        ...         ...        ...                ...   
81           Nicholls       1231  149.097395 -35.188480           750000.0   
82           O'Connor       1271  149.114719 -35.256153           750000.0   
83        Oaks Estate       1301  149.227164 -35.340004                NaN   
84              Oxley       1331  149.078932 -35.409159           476000.0   
85               Page       1361  149.050005 -35.237591           470000.0   
86         Palmerston       1371  149.119479 -35.193262           483000.0   
87             Parkes       1391  149.136227 -35.296097                NaN   
88             Pearce       1401  149.083778 -35.363060           650000.0   
89            Phillip       1411  149.091501 -35.348131           468000.0   
90           Red Hill       1451  149.120298 -35.332757          1075000.0   
91               Reid       1461  149.141205 -35.283233           930000.0   
92         Richardson       1471  149.108567 -35.426362           443000.0   
93             Rivett       1491  149.036776 -35.347684           490000.0   
94            Russell       3051  149.149682 -35.298164                NaN   
95            Scullin       1521  149.039432 -35.234771           465000.0   
96             Spence       1561  149.065484 -35.200341           473000.0   
97           Stirling       1581  149.049472 -35.349378           649000.0   
98          Symonston       1631  149.157218 -35.346989                NaN   
99             Tharwa       1661  149.064344 -35.508979                NaN   
100          Theodore       1671  149

In [58]:
order = [2,1,8,6,5,4,3,7,9,0]
canberra_afinal = pd.merge(canberra_house,suburb_venues,on = ['Suburb'], how = 'inner')
canberra_afinal['Cluster Labels'] = pd.Categorical(canberra_afinal['Cluster Labels'], order)
canberra_final = canberra_afinal.sort_values(['Cluster Labels', '5 yr Growth (%)', 'Median Price 2019'], ascending = [True, False, True])
canberra_final.head(10)

Suburb  Domain ID   Longitude   Latitude  Median Price 2014  \
1          Cook        441  149.066321 -35.260416           529000.0   
4   Narrabundah       1211  149.148882 -35.335116           660000.0   
5         Oxley       1331  149.078932 -35.409159           476000.0   
6      Franklin       2301  149.143494 -35.197892           530000.0   
8       Hackett        761  149.162325 -35.250558           663000.0   
11       Downer        541  149.143269 -35.242505           605000.0   
12      Dickson        531  149.140231 -35.254024           650000.0   
13       Farrer        611  149.102349 -35.376256           650000.0   
17        Lyons       1071  149.074145 -35.340090           615000.0   
20    Gungahlin        751  149.137635 -35.186028           542000.0   

    Median Price 2019  5 yr Growth (%) Cluster Labels 1st Most Common Venue  \
1            765000.0           144.61              2                  Café   
4            920000.0           139.39              2                 Hotel   
5            655000.0           137.61              2             Locksmith   
6            725000.0           136.79              2        Ice Cream Shop   
8            899000.0           135.60              2           Bus Station   
11           816000.0           134.88              2                  Café   
12           875000.0           134.62              2                  Café   
13           875000.0           134.62              2          Burger Joint   
17           820000.0           133.33              2                  Café   
20           714000.0           131.73              2                  Café   

   2nd Most Common Venue 3rd Most Common Venue    4th Most Common Venue  \
1                    Gym                   Bar               Sports Bar   
4            Golf Course                 Motel                      Gym   
5                Dog Run       Thai Restaurant                Gastropub   
6            Bus Station                   Spa  Health & Beauty Service   
8                   Café       Thai Restaurant            Grocery Store   
11           Coffee Shop        Sandwich Place          Thai Restaurant   
12    Chinese Restaurant      Asian Restaurant          Thai Restaurant   
13           Rugby Pitch                  Café            Grocery Store   
17            Sports Bar    Italian Restaurant             Skating Rink   
20           Supermarket           Coffee Shop                      Gym   

            5th Most Common Venue 6th Most Common Venue  \
1                     Pizza Place           Supermarket   
4               German Restaurant       Thai Restaurant   
5                     Yoga Studio           Film Studio   
6                            Café                  Park   
8                           Trail           Yoga Studio   
11                           Pool      Department Store   
12  Vegetarian / Vegan Restaurant                 Hotel   
13                    Film Studio     French Restaurant   
17           Gym / Fitness Center         Grocery Store   
20                 Ice Cream Shop       Thai Restaurant   

      7th Most Common Venue      8th Most Common Venue  \
1   Fruit & Vegetable Store               Liquor Store   
4               Supermarket   Mediterranean Restaurant   
5                  Fountain           Football Stadium   
6             Grocery Store            Thai Restaurant   
8       Filipino Restaurant           Football Stadium   
11                   Hostel  Middle Eastern Restaurant   
12        Korean Restaurant      Vietnamese Restaurant   
13                 Fountain           Football Stadium   
17      Filipino Restaurant           Football Stadium   
20         Department Store                        Pub   

            9th Most Common Venue 10th Most Common Venue  
1                   Grocery Store            Flea Market  
4                            Café       Baseball Stadium  
5                      Food Truck             Food Cou

In [297]:
# Cluster labels are not integers so something's wrong
#print('Cluster labels: {}'.format(canberra_suburbs['Cluster Labels'].unique()))
# Labels include NaN values for postcodes where no 4square results were returned
# therefore could either drop these postcodes or
# set NaN values to another integer value, e.g. 99, to include
# all postcodes in output but clearly indicating no venue results
# dropping rows with no venues
#canberra_suburbs.dropna(axis = 0, how ='any', inplace = True)
#canberra_suburbs=canberra_suburbs.astype({'Cluster Labels': int})
#print('Cleaned cluster labels: {}'.format(canberra_suburbs['Cluster Labels'].unique()))
#canberra_suburbs.head()

In [73]:
canberra_final.loc[canberra_final['Suburb'] == 'Kingston']

Empty DataFrame
Columns: [Suburb, Domain ID, Longitude, Latitude, Median Price 2014, Median Price 2019, 5 yr Growth (%), Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue, 8th Most Common Venue, 9th Most Common Venue, 10th Most Common Venue]
Index: []

### Create map

In [56]:
# Canberra map intial centre
c_lat, c_long = -35.2930556 , 149.126944

# create map
map_clusters = folium.Map(width=1000,height=1000,location=[c_lat, c_long], zoom_start=10)

# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(canberra_suburbs['Latitude'], canberra_suburbs['Longitude'], canberra_suburbs['Suburb'], canberra_suburbs['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Create final map

In [74]:
# create map
map_final = folium.Map(width=970,height=590,location=[c_lat, c_long], zoom_start=11)
# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(canberra_final['Latitude'], canberra_final['Longitude'], canberra_final['Suburb'], canberra_final['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_final)
       
map_final